include/index_factory.h

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "index.h"
+#include "common_includes.h"
+
+namespace diskann
+{
+template <typename T, typename TagT, typename LabelT> class MemoryIndex;
+template <typename T, typename LabelT> class DiskIndex;
+enum BuildType
+{
+    MEMORY,
+    DISK,
+    SSD
+};
+
+enum LoadStoreStratagy
+{
+};
+
+struct IndexConfig
+{
+    BuildType build_type;
+    Metric metric;
+    Parameters buildParams;
+    LoadStoreStratagy load_store_stratagy; // load and store stratagy when we have abstract data or abstract graph
+
+    bool filtered_build;
+
+    bool enable_tags = false;
+    bool dynamic_index = false;
+    bool pq_dist_build = false;
+    size_t num_pq_chunks = 0;
+    bool use_opq = false;
+    size_t num_frozen_pts = 0;
+    bool concurrent_consolidate = false;
+    bool load_on_build = false;
+
+    std::string data_type;
+    std::string label_type;
+    std::string tag_type;
+};
+
+struct BuildParams
+{
+    uint32_t R;
```

Comment:
opening discussion here: what say everyone to using more descriptive names for these parameters?  I took a stab at naming them in #300 for the old way of providing params, but I'm open to name changes from the ones I picked. 

I like single char variables in whitepapers, but I have a strong dislike for them when I'm trying to process the code I'm reading, especially if I'm not super familiar.

Reply:
I agree, and this was created with just POC. There is so much that we can improve here. First off reading those vars from a yml config file would be great.  Then as your #300 PR can be very helpful here.  I am planning to add yml-cpp library to do the config reading.